File: tmdb-core/src/main/java/dev/reuss/tmdb/domain/search/query/SearchMovieQuery.java
package dev.reuss.tmdb.domain.search.query;

import dev.reuss.tmdb.core.http.QueryParams;
import dev.reuss.tmdb.domain.search.SearchQuery;
import dev.reuss.tmdb.query.PagedQuery;
import dev.reuss.tmdb.query.QueryValidation;
import dev.reuss.tmdb.value.language.Language;
import dev.reuss.tmdb.value.region.Region;

/**
 * Query parameters for searching TMDB movies.
 */
public final class SearchMovieQuery implements SearchQuery, PagedQuery<SearchMovieQuery> {

    private final String query;
    private Boolean includeAdult;
    private Language language;
    private String primaryReleaseYear;
    private Integer page;
    private Region region;
    private String year;

    private SearchMovieQuery(String query) {
        this.query = QueryValidation.requireNotBlank(query, "Search query");
    }

    public static SearchMovieQuery of(String query) {
        return new SearchMovieQuery(query);
    }

    @Override
    public String query() {
        return query;
    }

    public SearchMovieQuery includeAdult(Boolean includeAdult) {
        this.includeAdult = includeAdult;
        return this;
    }

    public SearchMovieQuery language(Language language) {
        this.language = language;
        return this;
    }

    public SearchMovieQuery primaryReleaseYear(String primaryReleaseYear) {
        QueryValidation.validateYear(primaryReleaseYear, "Primary release year");
        this.primaryReleaseYear = primaryReleaseYear;
        return this;
    }

    public SearchMovieQuery page(Integer page) {
        QueryValidation.validatePage(page);
        this.page = page;
        return this;
    }

    public SearchMovieQuery region(Region region) {
        this.region = region;
        return this;
    }

    public SearchMovieQuery year(String year) {
        QueryValidation.validateYear(year, "Year");
        this.year = year;
        return this;
    }

    @Override
    public QueryParams toQueryParams() {
        return QueryParams.create()
                .add("query", query)
                .add("include_adult", includeAdult)
                .add("language", language)
                .add("primary_release_year", primaryReleaseYear)
                .add("page", page)
                .add("region", region)
                .add("year", year);
    }
}

File: tmdb-core/src/main/kotlin/dev/reuss/tmdb/domain/search/query/SearchMovieQuery.kt
package dev.reuss.tmdb.domain.search.query

import dev.reuss.tmdb.core.http.QueryParams
import dev.reuss.tmdb.query.PagedQuery
import dev.reuss.tmdb.query.QueryValidation
import dev.reuss.tmdb.value.language.Language
import dev.reuss.tmdb.value.region.Region

/**
 * Query parameters for searching TMDB movies.
 *
 * The search text is required and must not be blank. Optional filters are
 * adult content inclusion, language, page, region, year and primary release
 * year. Page and year filters are validated before serialization.
 */
class SearchMovieQuery private constructor(
    override val query: String
) : SearchQuery, PagedQuery<SearchMovieQuery> {

    private var includeAdult: Boolean? = null
    private var language: Language? = null
    private var primaryReleaseYear: Int? = null
    private var page: Int? = null
    private var region: Region? = null
    private var year: Int? = null

    fun includeAdult(value: Boolean?) = apply {
        includeAdult = value
    }

    fun language(value: Language?) = apply {
        language = value
    }

    fun primaryReleaseYear(value: Int) = apply {
        QueryValidation.validateYear(value, "Primary release year")
        primaryReleaseYear = value
    }

    override fun page(page: Int?) = apply {
        QueryValidation.validatePage(page)
        this.page = page
    }

    fun region(value: Region?) = apply {
        region = value
    }

    fun year(value: Int) = apply {
        QueryValidation.validateYear(value, "Year")
        year = value
    }

    override fun toQueryParams(): QueryParams =
        QueryParams.create()
            .add("query", query)
            .add("include_adult", includeAdult)
            .add("language", language?.value)
            .add("primary_release_year", primaryReleaseYear)
            .add("page", page)
            .add("region", region?.value)
            .add("year", year)

    companion object {

        @JvmStatic
        fun of(query: String): SearchMovieQuery =
            SearchMovieQuery(
                QueryValidation.requireNotBlank(query, "Search query")
            )
    }
}